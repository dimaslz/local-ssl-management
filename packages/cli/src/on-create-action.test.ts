import consola from "consola";
import crypto from "crypto";
import fs from "fs";
import shell from "shelljs";

import onCreateAction from "./on-create-action";

vi.mock("./list-container");

describe("On create action", () => {
  beforeEach(() => {
    vi.spyOn(fs, "readFileSync").mockReturnValue("[]");
    vi.spyOn(fs, "existsSync").mockReturnValueOnce(true);
  });

  describe("failures", () => {
    test("can not create with invalid domain", () => {
      const domain = "wrong.domain";
      const port = "3000";

      expect(() => {
        onCreateAction(domain, { port });
      }).toThrow();

      expect(consola.error).toBeCalledWith(
        new Error("Domain (https://wrong.domain) format is not valid"),
      );
      expect(shell.exit).toHaveBeenCalledWith(1);
    });

    test("can not create with invalid port", () => {
      const domain = "some-domain.com";
      const port = "666";

      expect(() => {
        onCreateAction(domain, { port });
      }).toThrow();

      expect(consola.error).toBeCalledWith(
        new Error(
          "Port (--port <port>) should be into the range 1025 to 65535",
        ),
      );
      expect(shell.exit).toHaveBeenCalledWith(1);
    });

    test("can not create duplicated location", () => {
      const domain = "some-domain.com";
      const port = "3000";
      const location = "/app-name";

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      vi.spyOn(fs, "readdirSync").mockImplementationOnce((): any[] => [
        "some-domain.com-cert.pem",
        "some-domain.com-key.pem",
      ]);

      vi.spyOn(fs, "readFileSync").mockReturnValueOnce(
        JSON.stringify([
          {
            id: "57ef8e87-7547-4bf5-a6a9-3dad7ffebe9e",
            domain: "some-domain.com",
            services: [
              {
                id: "a4eb6238-f1ab-418e-b910-29d73192637a",
                port: "4000",
                location: "/",
              },
              {
                id: "a4eb6238-f1ab-418e-b910-29d73192637a",
                port: "3000",
                location: "/app-name",
              },
            ],
          },
        ]),
      );

      expect(() => {
        onCreateAction(domain, { port, location });
      }).toThrow();

      expect(consola.error).toBeCalledWith(
        new Error('Location "/app-name" already exists'),
      );
      expect(shell.exit).toHaveBeenCalledWith(1);
    });

    test("can not create duplicated domain", () => {
      const domain = "some-domain.com";

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      vi.spyOn(fs, "readdirSync").mockImplementationOnce((): any[] => [
        "some-domain.com-cert.pem",
        "some-domain.com-key.pem",
      ]);

      vi.spyOn(fs, "readFileSync").mockReturnValueOnce(
        JSON.stringify([
          {
            id: "57ef8e87-7547-4bf5-a6a9-3dad7ffebe9e",
            domain: "some-domain.com",
            services: [
              {
                id: "a4eb6238-f1ab-418e-b910-29d73192637a",
                port: "4000",
                location: "/",
              },
              {
                id: "a4eb6238-f1ab-418e-b910-29d73192637a",
                port: "3000",
                location: "/app-name",
              },
            ],
          },
        ]),
      );

      expect(() => {
        onCreateAction(domain, {});
      }).toThrow();

      expect(consola.error).toBeCalledWith(
        new Error(
          'Domain "some-domain.com" already created with the default location "/"',
        ),
      );
      expect(shell.exit).toHaveBeenCalledWith(1);
    });
  });

  describe("success", () => {
    beforeEach(() => {
      vi.spyOn(fs, "mkdirSync");
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      vi.spyOn(fs, "readdirSync").mockImplementationOnce((): any[] => [
        "some-domain.com-cert.pem",
        "some-domain.com-key.pem",
      ]);

      vi.spyOn(shell, "exec")
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        .mockImplementationOnce((): any => "")
        .mockImplementationOnce(
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          (): any => {
            return { stdout: 200 };
          },
        );
    });

    describe("single service", () => {
      test("domain config created sucessfully", () => {
        crypto.randomUUID = vi.fn(() => "48d1a85c-377a-40ef-8a82-d1405f7a074f");

        const domain = "some-domain.com";
        const port = "3000";

        onCreateAction(domain, { port });

        expect(shell.exec).toBeCalledTimes(2);
        expect(shell.exec).nthCalledWith(
          1,
          "NAME=local-ssl-management && docker rm -f $NAME && docker rmi -f $NAME && docker build --no-cache -t $NAME /Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management && docker run --name $NAME -p 80:80 -p 443:443 -d $NAME",
          { silent: true },
        );
        expect(shell.exec).nthCalledWith(
          2,
          'curl -s -o /dev/null -w "%{http_code}" https://some-domain.com',
          { silent: true },
        );

        expect(fs.readFileSync).toBeCalledTimes(1);
        expect(fs.readFileSync).toHaveBeenNthCalledWith(
          1,
          expect.stringMatching(/.local-ssl-management\/config.json/),
          { encoding: "utf8" },
        );

        expect(fs.existsSync).toHaveBeenCalledWith(
          expect.stringMatching(/.local-ssl-management\/ssl/),
        );
        expect(fs.mkdirSync).not.toHaveBeenCalled();

        expect(fs.writeFileSync).toHaveBeenNthCalledWith(
          3,
          expect.stringMatching(/.local-ssl-management\/config.json/),
          JSON.stringify(
            [
              {
                id: "48d1a85c-377a-40ef-8a82-d1405f7a074f",
                domain: "some-domain.com",
                services: [
                  {
                    id: "48d1a85c-377a-40ef-8a82-d1405f7a074f",
                    location: "/",
                    port: "3000",
                  },
                ],
              },
            ],
            null,
            2,
          ),
        );
      });

      test("domain config created sucessfully (does not exists /ssl)", () => {
        vi.spyOn(fs, "readFileSync").mockReturnValue(JSON.stringify([]));
        vi.spyOn(fs, "existsSync").mockReturnValue(false);

        vi.spyOn(crypto, "randomUUID")
          .mockImplementationOnce(
            vi.fn(() => "48d1a85c-377a-40ef-8a82-d1405f7a074f"),
          )
          .mockImplementationOnce(
            vi.fn(() => "63add84e-b600-4f0f-9e1a-7ea85cfa2d06"),
          );

        const domain = "some-domain.com";
        const port = "3000";

        onCreateAction(domain, { port });

        expect(shell.exec).toBeCalledTimes(2);
        expect(shell.exec).nthCalledWith(
          1,
          "NAME=local-ssl-management && docker rm -f $NAME && docker rmi -f $NAME && docker build --no-cache -t $NAME /Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management && docker run --name $NAME -p 80:80 -p 443:443 -d $NAME",
          { silent: true },
        );
        expect(shell.exec).nthCalledWith(
          2,
          'curl -s -o /dev/null -w "%{http_code}" https://some-domain.com',
          { silent: true },
        );

        expect(fs.readFileSync).toHaveBeenNthCalledWith(
          1,
          expect.stringMatching(/.local-ssl-management\/config.json/),
          { encoding: "utf8" },
        );
        expect(fs.existsSync).toHaveBeenCalledWith(
          expect.stringMatching(/.local-ssl-management\/ssl/),
        );
        expect(fs.mkdirSync).toHaveBeenCalled();

        expect(fs.writeFileSync).toHaveBeenCalledWith(
          expect.stringMatching(/.local-ssl-management\/config.json/),
          JSON.stringify(
            [
              {
                id: "48d1a85c-377a-40ef-8a82-d1405f7a074f",
                domain: "some-domain.com",
                services: [
                  {
                    id: "63add84e-b600-4f0f-9e1a-7ea85cfa2d06",
                    location: "/",
                    port: "3000",
                  },
                ],
              },
            ],
            null,
            2,
          ),
        );
      });

      test("domain config already exists", () => {
        vi.spyOn(fs, "readFileSync").mockReturnValue(
          JSON.stringify(
            [
              {
                id: "48d1a85c-377a-40ef-8a82-d1405f7a074f",
                domain: "some-domain.com",
                services: [
                  {
                    id: "8145baf2-6e39-453a-a842-9b4a4f576003",
                    port: "3000",
                    location: "/",
                  },
                ],
              },
            ],
            null,
            2,
          ),
        );

        const domain = "some-domain.com";
        const port = "3000";

        expect(() => {
          onCreateAction(domain, { port });
        }).toThrow();

        expect(consola.error).toBeCalledWith(
          new Error(
            'Domain "some-domain.com" already created with the default location "/"',
          ),
        );
        expect(shell.exit).toHaveBeenCalledWith(1);
      });
    });

    describe("multiple services", () => {
      test("domain config created sucessfully", () => {
        vi.spyOn(fs, "readFileSync").mockReturnValue(
          JSON.stringify(
            [
              {
                id: "54a04fef-f263-4eab-a17a-016c76986160",
                domain: "some-domain.com",
                services: [
                  {
                    id: "ac8a5cb1-a331-4888-8b6f-d99fc277a1ca",
                    port: "3000",
                    location: "/",
                  },
                ],
              },
            ],
            null,
            2,
          ),
        );

        vi.spyOn(fs, "existsSync").mockReturnValue(true);

        vi.spyOn(crypto, "randomUUID").mockImplementationOnce(
          vi.fn(() => "48d1a85c-377a-40ef-8a82-d1405f7a074f"),
        );

        const domain = "some-domain.com";
        const port = "4000";
        const location = "/app-name";

        onCreateAction(domain, { port, location });

        expect(shell.exec).toBeCalledTimes(2);
        expect(shell.exec).nthCalledWith(
          1,
          "NAME=local-ssl-management && docker rm -f $NAME && docker rmi -f $NAME && docker build --no-cache -t $NAME /Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management && docker run --name $NAME -p 80:80 -p 443:443 -d $NAME",
          { silent: true },
        );
        expect(shell.exec).nthCalledWith(
          2,
          'curl -s -o /dev/null -w "%{http_code}" https://some-domain.com',
          { silent: true },
        );

        expect(fs.readFileSync).toHaveBeenNthCalledWith(
          1,
          expect.stringMatching(/.local-ssl-management\/config.json/),
          { encoding: "utf8" },
        );
        expect(fs.existsSync).toHaveBeenCalledWith(
          expect.stringMatching(/.local-ssl-management\/ssl/),
        );
        expect(fs.mkdirSync).not.toHaveBeenCalled();

        expect(fs.writeFileSync).toBeCalledTimes(3);
        expect(fs.writeFileSync).toMatchSnapshot();

        expect(fs.writeFileSync).toHaveBeenNthCalledWith(
          3,
          expect.stringMatching(/.local-ssl-management\/config.json/),
          JSON.stringify(
            [
              {
                id: "54a04fef-f263-4eab-a17a-016c76986160",
                domain: "some-domain.com",
                services: [
                  {
                    id: "ac8a5cb1-a331-4888-8b6f-d99fc277a1ca",
                    port: "3000",
                    location: "/",
                  },
                  {
                    id: "48d1a85c-377a-40ef-8a82-d1405f7a074f",
                    location: "/app-name",
                    port: "4000",
                  },
                ],
              },
            ],
            null,
            2,
          ),
        );
      });

      test("domain config created sucessfully (does not exists /ssl)", () => {
        vi.spyOn(fs, "readFileSync").mockReturnValue(JSON.stringify([]));
        vi.spyOn(fs, "existsSync").mockReturnValue(false);

        vi.spyOn(crypto, "randomUUID")
          .mockImplementationOnce(
            vi.fn(() => "48d1a85c-377a-40ef-8a82-d1405f7a074f"),
          )
          .mockImplementationOnce(
            vi.fn(() => "63add84e-b600-4f0f-9e1a-7ea85cfa2d06"),
          );

        const domain = "some-domain.com";
        const port = "3000";

        onCreateAction(domain, { port });

        expect(shell.exec).toBeCalledTimes(2);
        expect(shell.exec).nthCalledWith(
          1,
          "NAME=local-ssl-management && docker rm -f $NAME && docker rmi -f $NAME && docker build --no-cache -t $NAME /Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management && docker run --name $NAME -p 80:80 -p 443:443 -d $NAME",
          { silent: true },
        );
        expect(shell.exec).nthCalledWith(
          2,
          'curl -s -o /dev/null -w "%{http_code}" https://some-domain.com',
          { silent: true },
        );

        expect(fs.readFileSync).toHaveBeenNthCalledWith(
          1,
          expect.stringMatching(/.local-ssl-management\/config.json/),
          { encoding: "utf8" },
        );
        expect(fs.existsSync).toHaveBeenCalledWith(
          expect.stringMatching(/.local-ssl-management\/ssl/),
        );
        expect(fs.mkdirSync).toHaveBeenCalled();

        expect(fs.writeFileSync).toBeCalledTimes(3);
        expect(fs.writeFileSync).toMatchSnapshot();

        expect(fs.writeFileSync).toHaveBeenNthCalledWith(
          3,
          expect.stringMatching(/.local-ssl-management\/config.json/),
          JSON.stringify(
            [
              {
                id: "48d1a85c-377a-40ef-8a82-d1405f7a074f",
                domain: "some-domain.com",
                services: [
                  {
                    id: "63add84e-b600-4f0f-9e1a-7ea85cfa2d06",
                    location: "/",
                    port: "3000",
                  },
                ],
              },
            ],
            null,
            2,
          ),
        );
      });

      test("domain config already exists", () => {
        vi.spyOn(fs, "readFileSync").mockReturnValue(
          JSON.stringify(
            [
              {
                id: "48d1a85c-377a-40ef-8a82-d1405f7a074f",
                domain: "some-domain.com",
                services: [
                  {
                    id: "8145baf2-6e39-453a-a842-9b4a4f576003",
                    port: "3000",
                    location: "/",
                  },
                ],
              },
            ],
            null,
            2,
          ),
        );

        const domain = "some-domain.com";
        const port = "3000";

        expect(() => {
          onCreateAction(domain, { port });
        }).toThrow();

        expect(consola.error).toBeCalledWith(
          new Error(
            'Domain "some-domain.com" already created with the default location "/"',
          ),
        );
        expect(shell.exit).toHaveBeenCalledWith(1);
      });
    });
  });
});
