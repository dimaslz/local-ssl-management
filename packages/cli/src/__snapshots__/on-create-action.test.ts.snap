// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`On create action > success > multiple services > domain config created sucessfully (does not exists /ssl) 1`] = `
[MockFunction writeFileSync] {
  "calls": [
    [
      "/Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management/nginx.conf",
      "user  nginx;
worker_processes  20;

error_log		/var/log/nginx/error.log warn;
pid					/var/run/nginx.pid;

events {
	worker_connections  1024;
}

http {
		server {
				listen              443 ssl;
				server_name         _;
				ssl_certificate     /etc/nginx/localhost-cert.pem;
				ssl_certificate_key /etc/nginx/localhost-key.pem;
				location / {
						root  /var/www/html;
				}
		}

		server {
				listen 80 default_server;
				server_name         _;

				include       /etc/nginx/mime.types;
				default_type  application/octet-stream;

				location / {
						root  /var/www/html;
				}
		}

		server {
			listen	443 ssl;

			autoindex off;

			access_log  /var/log/nginx/some-domain.com.access.log;
			error_log   /var/log/nginx/some-domain.com.error.log;

			server_tokens off;
			server_name some-domain.com;

			ssl_certificate     /etc/nginx/some-domain.com-cert.pem;
			ssl_certificate_key /etc/nginx/some-domain.com-key.pem;

			gzip_static on;

			location ~ ^/(/?)(.*) {
				gzip on;
				gzip_disable "msie6";
				gzip_vary on;
				gzip_proxied any;
				gzip_comp_level 6;
				gzip_buffers 16 8k;
				gzip_http_version 1.1;
				gzip_min_length 256;
				gzip_types text/plain text/css application/json application/x-javascript application/javascript text/xml application/xml application/xml+rss text/javascript application/vnd.ms-fontobject application/x-font-ttf font/opentype image/svg+xml image/x-icon;
				proxy_pass http://11.22.33.445:3000/$2;
				proxy_redirect off;
				proxy_http_version 1.1;
				proxy_cache_bypass $http_upgrade;
				proxy_set_header Upgrade $http_upgrade;
				proxy_set_header Connection 'upgrade';
				proxy_set_header Host $host;
				proxy_set_header 'Access-Control-Allow-Origin' '*';
				proxy_set_header X-Real-IP $remote_addr;
				proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
				proxy_set_header 'Cache-Control' 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
				expires off;
		}
	}
}",
    ],
    [
      "/Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management/Dockerfile",
      "FROM nginx

# RUN rm -f /etc/nginx/conf.d/default.conf

# WORKDIR /var/www/html
# COPY index.html /var/www/html
# RUN chmod 755 /var/www/html/index.html

COPY nginx.conf /etc/nginx/conf.d/

COPY ssl/some-domain.com-key.pem /etc/nginx/
COPY ssl/some-domain.com-cert.pem /etc/nginx/

COPY ssl/localhost-key.pem /etc/nginx/
COPY ssl/localhost-cert.pem /etc/nginx/

COPY nginx.conf /etc/nginx/

EXPOSE 80 443

CMD ["nginx", "-g", "daemon off;"]",
    ],
    [
      "/Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management/config.json",
      "[
  {
    "id": "48d1a85c-377a-40ef-8a82-d1405f7a074f",
    "domain": "some-domain.com",
    "services": [
      {
        "id": "da406b35-79b8-4909-9af1-07cfdcf37d00",
        "location": "/",
        "port": "3000"
      }
    ]
  }
]",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`On create action > success > multiple services > domain config created sucessfully 1`] = `
[MockFunction writeFileSync] {
  "calls": [
    [
      "/Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management/nginx.conf",
      "user  nginx;
worker_processes  20;

error_log		/var/log/nginx/error.log warn;
pid					/var/run/nginx.pid;

events {
	worker_connections  1024;
}

http {
		server {
				listen              443 ssl;
				server_name         _;
				ssl_certificate     /etc/nginx/localhost-cert.pem;
				ssl_certificate_key /etc/nginx/localhost-key.pem;
				location / {
						root  /var/www/html;
				}
		}

		server {
				listen 80 default_server;
				server_name         _;

				include       /etc/nginx/mime.types;
				default_type  application/octet-stream;

				location / {
						root  /var/www/html;
				}
		}

		server {
			listen	443 ssl;

			autoindex off;

			access_log  /var/log/nginx/some-domain.com.access.log;
			error_log   /var/log/nginx/some-domain.com.error.log;

			server_tokens off;
			server_name some-domain.com;

			ssl_certificate     /etc/nginx/some-domain.com-cert.pem;
			ssl_certificate_key /etc/nginx/some-domain.com-key.pem;

			gzip_static on;

			location ~ ^/(/?)(.*) {
				gzip on;
				gzip_disable "msie6";
				gzip_vary on;
				gzip_proxied any;
				gzip_comp_level 6;
				gzip_buffers 16 8k;
				gzip_http_version 1.1;
				gzip_min_length 256;
				gzip_types text/plain text/css application/json application/x-javascript application/javascript text/xml application/xml application/xml+rss text/javascript application/vnd.ms-fontobject application/x-font-ttf font/opentype image/svg+xml image/x-icon;
				proxy_pass http://11.22.33.445:3000/$2;
				proxy_redirect off;
				proxy_http_version 1.1;
				proxy_cache_bypass $http_upgrade;
				proxy_set_header Upgrade $http_upgrade;
				proxy_set_header Connection 'upgrade';
				proxy_set_header Host $host;
				proxy_set_header 'Access-Control-Allow-Origin' '*';
				proxy_set_header X-Real-IP $remote_addr;
				proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
				proxy_set_header 'Cache-Control' 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
				expires off;
		}

location ~ ^/app-name(/?)(.*) {
				gzip on;
				gzip_disable "msie6";
				gzip_vary on;
				gzip_proxied any;
				gzip_comp_level 6;
				gzip_buffers 16 8k;
				gzip_http_version 1.1;
				gzip_min_length 256;
				gzip_types text/plain text/css application/json application/x-javascript application/javascript text/xml application/xml application/xml+rss text/javascript application/vnd.ms-fontobject application/x-font-ttf font/opentype image/svg+xml image/x-icon;
				proxy_pass http://11.22.33.445:4000/$2;
				proxy_redirect off;
				proxy_http_version 1.1;
				proxy_cache_bypass $http_upgrade;
				proxy_set_header Upgrade $http_upgrade;
				proxy_set_header Connection 'upgrade';
				proxy_set_header Host $host;
				proxy_set_header 'Access-Control-Allow-Origin' '*';
				proxy_set_header X-Real-IP $remote_addr;
				proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
				proxy_set_header 'Cache-Control' 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
				expires off;
		}
	}
}",
    ],
    [
      "/Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management/Dockerfile",
      "FROM nginx

# RUN rm -f /etc/nginx/conf.d/default.conf

# WORKDIR /var/www/html
# COPY index.html /var/www/html
# RUN chmod 755 /var/www/html/index.html

COPY nginx.conf /etc/nginx/conf.d/

COPY ssl/some-domain.com-key.pem /etc/nginx/
COPY ssl/some-domain.com-cert.pem /etc/nginx/

COPY ssl/localhost-key.pem /etc/nginx/
COPY ssl/localhost-cert.pem /etc/nginx/

COPY nginx.conf /etc/nginx/

EXPOSE 80 443

CMD ["nginx", "-g", "daemon off;"]",
    ],
    [
      "/Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management/config.json",
      "[
  {
    "id": "54a04fef-f263-4eab-a17a-016c76986160",
    "domain": "some-domain.com",
    "services": [
      {
        "id": "ac8a5cb1-a331-4888-8b6f-d99fc277a1ca",
        "port": "3000",
        "location": "/"
      },
      {
        "id": "48d1a85c-377a-40ef-8a82-d1405f7a074f",
        "location": "/app-name",
        "port": "4000"
      }
    ]
  }
]",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`On create action > success > single service > domain config created sucessfully (does not exists /ssl) 1`] = `
[MockFunction writeFileSync] {
  "calls": [
    [
      "/Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management/nginx.conf",
      "user  nginx;
worker_processes  20;

error_log		/var/log/nginx/error.log warn;
pid					/var/run/nginx.pid;

events {
	worker_connections  1024;
}

http {
		server {
				listen              443 ssl;
				server_name         _;
				ssl_certificate     /etc/nginx/localhost-cert.pem;
				ssl_certificate_key /etc/nginx/localhost-key.pem;
				location / {
						root  /var/www/html;
				}
		}

		server {
				listen 80 default_server;
				server_name         _;

				include       /etc/nginx/mime.types;
				default_type  application/octet-stream;

				location / {
						root  /var/www/html;
				}
		}

		server {
			listen	443 ssl;

			autoindex off;

			access_log  /var/log/nginx/some-domain.com.access.log;
			error_log   /var/log/nginx/some-domain.com.error.log;

			server_tokens off;
			server_name some-domain.com;

			ssl_certificate     /etc/nginx/some-domain.com-cert.pem;
			ssl_certificate_key /etc/nginx/some-domain.com-key.pem;

			gzip_static on;

			location ~ ^/(/?)(.*) {
				gzip on;
				gzip_disable "msie6";
				gzip_vary on;
				gzip_proxied any;
				gzip_comp_level 6;
				gzip_buffers 16 8k;
				gzip_http_version 1.1;
				gzip_min_length 256;
				gzip_types text/plain text/css application/json application/x-javascript application/javascript text/xml application/xml application/xml+rss text/javascript application/vnd.ms-fontobject application/x-font-ttf font/opentype image/svg+xml image/x-icon;
				proxy_pass http://11.22.33.445:3000/$2;
				proxy_redirect off;
				proxy_http_version 1.1;
				proxy_cache_bypass $http_upgrade;
				proxy_set_header Upgrade $http_upgrade;
				proxy_set_header Connection 'upgrade';
				proxy_set_header Host $host;
				proxy_set_header 'Access-Control-Allow-Origin' '*';
				proxy_set_header X-Real-IP $remote_addr;
				proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
				proxy_set_header 'Cache-Control' 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
				expires off;
		}
	}
}",
    ],
    [
      "/Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management/Dockerfile",
      "FROM nginx

# RUN rm -f /etc/nginx/conf.d/default.conf

# WORKDIR /var/www/html
# COPY index.html /var/www/html
# RUN chmod 755 /var/www/html/index.html

COPY nginx.conf /etc/nginx/conf.d/

COPY ssl/some-domain.com-key.pem /etc/nginx/
COPY ssl/some-domain.com-cert.pem /etc/nginx/

COPY ssl/localhost-key.pem /etc/nginx/
COPY ssl/localhost-cert.pem /etc/nginx/

COPY nginx.conf /etc/nginx/

EXPOSE 80 443

CMD ["nginx", "-g", "daemon off;"]",
    ],
    [
      "/Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management/config.json",
      "[
  {
    "id": "48d1a85c-377a-40ef-8a82-d1405f7a074f",
    "domain": "some-domain.com",
    "services": [
      {
        "id": "da406b35-79b8-4909-9af1-07cfdcf37d00",
        "location": "/",
        "port": "3000"
      }
    ]
  }
]",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`On create action > success > single service > domain config created sucessfully 1`] = `
[MockFunction writeFileSync] {
  "calls": [
    [
      "/Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management/nginx.conf",
      "user  nginx;
worker_processes  20;

error_log		/var/log/nginx/error.log warn;
pid					/var/run/nginx.pid;

events {
	worker_connections  1024;
}

http {
		server {
				listen              443 ssl;
				server_name         _;
				ssl_certificate     /etc/nginx/localhost-cert.pem;
				ssl_certificate_key /etc/nginx/localhost-key.pem;
				location / {
						root  /var/www/html;
				}
		}

		server {
				listen 80 default_server;
				server_name         _;

				include       /etc/nginx/mime.types;
				default_type  application/octet-stream;

				location / {
						root  /var/www/html;
				}
		}

		server {
			listen	443 ssl;

			autoindex off;

			access_log  /var/log/nginx/some-domain.com.access.log;
			error_log   /var/log/nginx/some-domain.com.error.log;

			server_tokens off;
			server_name some-domain.com;

			ssl_certificate     /etc/nginx/some-domain.com-cert.pem;
			ssl_certificate_key /etc/nginx/some-domain.com-key.pem;

			gzip_static on;

			location ~ ^/(/?)(.*) {
				gzip on;
				gzip_disable "msie6";
				gzip_vary on;
				gzip_proxied any;
				gzip_comp_level 6;
				gzip_buffers 16 8k;
				gzip_http_version 1.1;
				gzip_min_length 256;
				gzip_types text/plain text/css application/json application/x-javascript application/javascript text/xml application/xml application/xml+rss text/javascript application/vnd.ms-fontobject application/x-font-ttf font/opentype image/svg+xml image/x-icon;
				proxy_pass http://11.22.33.445:3000/$2;
				proxy_redirect off;
				proxy_http_version 1.1;
				proxy_cache_bypass $http_upgrade;
				proxy_set_header Upgrade $http_upgrade;
				proxy_set_header Connection 'upgrade';
				proxy_set_header Host $host;
				proxy_set_header 'Access-Control-Allow-Origin' '*';
				proxy_set_header X-Real-IP $remote_addr;
				proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
				proxy_set_header 'Cache-Control' 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
				expires off;
		}
	}
}",
    ],
    [
      "/Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management/Dockerfile",
      "FROM nginx

# RUN rm -f /etc/nginx/conf.d/default.conf

# WORKDIR /var/www/html
# COPY index.html /var/www/html
# RUN chmod 755 /var/www/html/index.html

COPY nginx.conf /etc/nginx/conf.d/

COPY ssl/some-domain.com-key.pem /etc/nginx/
COPY ssl/some-domain.com-cert.pem /etc/nginx/

COPY ssl/localhost-key.pem /etc/nginx/
COPY ssl/localhost-cert.pem /etc/nginx/

COPY nginx.conf /etc/nginx/

EXPOSE 80 443

CMD ["nginx", "-g", "daemon off;"]",
    ],
    [
      "/Users/dimaslz/Development/local-ssl-management/packages/cli/src/.local-ssl-management/config.json",
      "[
  {
    "id": "48d1a85c-377a-40ef-8a82-d1405f7a074f",
    "domain": "some-domain.com",
    "services": [
      {
        "id": "da406b35-79b8-4909-9af1-07cfdcf37d00",
        "location": "/",
        "port": "3000"
      }
    ]
  }
]",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
